compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -340,6 +340,33 @@ impl<'ll, 'tcx> IntrinsicCallBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
                     self.const_i32(cache_type),
                 ])
             }
+            sym::carrying_mul_add => {
+                let uty = fn_args.type_at(0);
+                let (size, signed) = uty.int_size_and_signed(self.tcx);
+                assert!(!signed);
```

Comment:
Is there a particular reason why you exclude signedness here? I don't see anything particular here that depends on the integers being unsigned, since you could easily change `zext` into `intcast` to make this work like I did in my implementation: https://github.com/rust-lang/rust/pull/131566/files#diff-a762eadbc5da5c5fce9dff3c4a11e4ad720e6580767cc097d49e64264858aa61R463-R494

(not sure if this affects codegen, though, since I never tested that)

Reply:
Ah! Fair; I did manage to make it work in my purely library solution, but haven't since compared the two. I mostly just special-cased the signed versions.